Adding support for 3D operators

GeorgeR227 opened the pull request:
This PR is meant to lift features that work in 1D and 2D to also work in 3D. This includes automatic typing and operator overloading as well as operator replacement. It also adds new `:Form3` and `:DualForm3` types to allow these changes. Additional additions may be needed in the future to allow for full support.

Commit:
Added rules for infering and overloading 3D operators

GeorgeR227 committed:
Added wedge for 0-3 forms

lukem12345 committed:
Merge branch 'main' into gr/3d_rules

Copilot committed:
Fix 3D ThDEC typing aliases and docstring

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/c769e360-3940-4906-a5f4-021c9d9bbb49

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

Copilot committed:
Refactor 3D alias declarations for readability

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/c769e360-3940-4906-a5f4-021c9d9bbb49

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

Copilot committed:
Tighten ThDEC alias lists for 3D operators

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/c769e360-3940-4906-a5f4-021c9d9bbb49

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

Copilot committed:
Replace fragile positional ACSet tests with pprint(Term(deca_3d)) comparison

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/1b38a7c2-6ea6-49ca-9858-bbf9f45c09ce

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

Copilot committed:
Use triple-quoted string for readability in pprint comparison test

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/1b38a7c2-6ea6-49ca-9858-bbf9f45c09ce

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

lukem12345 commented on src/deca/deca_acset.jl:
```diff
@@ -121,9 +150,14 @@ op2_operators = [
 # use dispatch to choose the correct set of rules.
 function default_operators(dim)
-  @assert 1 <= dim <= 2
+  @assert 1 <= dim <= 3
   metric_free = vcat(op1_operators, op2_operators)
-  return vcat(metric_free, dim == 1 ? op1_1D_bound_operators : op1_2D_bound_operators)
+  metric_bound = @match dim begin
+    1 => op1_1D_bound_operators
+    2 => op1_2D_bound_operators
+    3 => op1_3D_bound_operators
+  end
+  return vcat(metric_free, metric_bound)
```
Known architecture choice due to properties of the exterior algebra.

lukem12345 commented on test/language.jl:
```diff
@@ -1230,6 +1230,116 @@ end
+R21   = ∧₃₀(P3, P0)
+R22   = ∧₀₃(P0, P3)
+"""
+  @test sprint((io, x) -> DiagrammaticEquations.pprint(io, x), Term(deca_3d)) == expected_pprint
```
@copilot Split the triple-quoted string on newlines and check for set equality with the sprint output, similarly split. Also add appropriate rewrite rules for the 3-Laplacian as noted in the other review comment.

Copilot committed:
Set equality for pprint test, add isForm3 predicate and Δ₃ rewrite rule

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/464856f1-0aa1-4907-a10a-92f546a30f8d

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>

Copilot committed:
Use issetequal for 3D pprint line-set assertion

Agent-Logs-Url: https://github.com/AlgebraicJulia/DiagrammaticEquations.jl/sessions/d334328d-d482-4fc4-a87d-c8ce4441817d

Co-authored-by: lukem12345 <70283489+lukem12345@users.noreply.github.com>
